### PR TITLE
Fix #174: remove `vop` from `M` commands

### LINKED ISSUE #174: Remove `vop` from the `M` commands
**Is your feature request related to a problem? Please describe.**
(to be worked on after merging #155): `vop` info is redundant as it can be converted to measurement planes and angle changes.
Modify them instead of adding`vop`, in e.g. `measure_pauli` `traspile(opt=True)`, etc. remove `vop` from measurement command definition.

we may deprecate `opt=True` transpile so we may want to do this after that.

### PR DESCRIPTION
**Pull request by thierry-martinez**
This commit removes the `vop` attribute from `M` commands, and adds a `clifford` method to `M` commands: this method takes a Clifford gate and returns a new `M` command where the plane, the angle, and the s- and t-domains have been updated, so as to be equivalent to apply the given Clifford gate prior to the measure.

Standardization, signal shifting and tensor network simulator have been fixed for planes other than XY.

### CODE REVIEW

File: graphix/command.py [hunk 1]
```diff
@@ -5,8 +5,10 @@
 import abc
 import enum
 
+import numpy as np
 from pydantic import BaseModel
 
+import graphix.clifford
```
Comment (shinich1) [inline graphix/command.py:11]: `from graphix.clifford import Clifford` ?

File: graphix/pattern.py [hunk 2]
```diff
@@ -2179,3 +2180,20 @@ def assert_permutation(original, user):
             node_set.remove(node)
         else:
             assert False, f"{node} appears twice"
+
+
+@dataclass
+class ExtractedSignal:
+    s_domain: list[int]
+    t_domain: list[int]
+    signal: list[int]
```
Comment (shinich1) [inline graphix/pattern.py:2189]: what's the `signal` of `ExtractedSignal`? is this different from domains?

**Comment by shinich1** [comment]
@thierry-martinez thanks! this looks really nice approach. examples rendered well. we might want to add `Clifford` to API docs page, `docs/source/data.rst` (I can do this after this is merged).
Can @EarlMilktea also take a quick look?

File: graphix/command.py [hunk 3]
```diff
@@ -51,7 +53,30 @@ class M(Command):
     angle: float = 0.0
     s_domain: list[Node] = []
     t_domain: list[Node] = []
-    vop: int = 0
+
+    def clifford(self, clifford: Clifford) -> M:
```
Comment (EarlMilktea) [inline graphix/command.py:57]: `Clifford` is unresolved.

File: graphix/command.py [hunk 4]
```diff
@@ -51,7 +53,30 @@ class M(Command):
...
+            if gate == graphix.clifford.I:
+                pass
+            elif gate == graphix.clifford.H:
+                t_domain, s_domain = s_domain, t_domain
+            elif gate == graphix.clifford.S:
+                t_domain = s_domain + t_domain
+            elif gate == graphix.clifford.Z:
+                pass
+            else:
+                assert False
```
Comment (EarlMilktea) [inline graphix/command.py:70]: `assert` can be bypassed. Please raise `RuntimeError` instead.

File: graphix/command.py [hunk 5]
```diff
@@ -51,7 +53,30 @@ class M(Command):
...
+        for gate in clifford.hsz:
+            if gate == graphix.clifford.I:
+                pass
+            elif gate == graphix.clifford.H:
+                t_domain, s_domain = s_domain, t_domain
+            elif gate == graphix.clifford.S:
+                t_domain = s_domain + t_domain
+            elif gate == graphix.clifford.Z:
+                pass
+            else:
```
Comment (EarlMilktea) [inline graphix/command.py:69]: May better to annotate types and use `typing.assert_never`.

File: graphix/command.py [hunk 6]
```diff
@@ -51,7 +53,30 @@ class M(Command):
...
+                assert False
+        update = graphix.pauli.MeasureUpdate.compute(self.plane, False, False, clifford)
+        result = M(
```
Comment (EarlMilktea) [inline graphix/command.py:72]: Please `return` directly.

File: graphix/command.py (hunk 1 above)
Comment (shinich1) [reply graphix/command.py:11]: `Clifford` used later

File: graphix/pattern.py [hunk 7]
```diff
@@ -13,9 +14,11 @@
 import graphix.pauli
 from graphix import command
 from graphix.clifford import CLIFFORD_CONJ, CLIFFORD_MEASURE, CLIFFORD_TO_QASM3
+from graphix.command import CommandKind
```
Comment (EarlMilktea) [inline graphix/pattern.py:17]: Unused.

File: graphix/pattern.py [hunk 8]
```diff
@@ -1421,6 +1417,16 @@ def to_qasm3(self, filename):
                 for line in cmd_to_qasm3(command):
                     file.write(line)
 
+    def copy(self):
```
Comment (EarlMilktea) [inline graphix/pattern.py:1420]: Please do not omit return type annotation.

File: graphix/pattern.py [hunk 9]
```diff
@@ -2179,3 +2180,20 @@ def assert_permutation(original, user):
...
+
+
+def extract_signal(plane: Plane, s_domain: list[int], t_domain: list[int]) -> ExtractedSignal:
+    if plane == Plane.XY:
+        return ExtractedSignal(s_domain=s_domain, t_domain=[], signal=t_domain)
+    if plane == Plane.XZ:
+        return ExtractedSignal(s_domain=[], t_domain=s_domain + t_domain, signal=s_domain)
+    if plane == Plane.YZ:
+        return ExtractedSignal(s_domain=[], t_domain=t_domain, signal=s_domain)
+    assert False
```
Comment (EarlMilktea) [inline graphix/pattern.py:2199]: Same as above.

File: graphix/sim/base_backend.py [hunk 10]
```diff
@@ -22,19 +22,19 @@ def perform_measure(
...
-        result = np.random.choice([0, 1])
+        result = rng.choice([0, 1])
         op_mat = op_mat_from_result(vec, result)
     state.evolve_single(op_mat, qubit)
     return result
 
 
 class Backend:
-    def __init__(self, pr_calc: bool = True):
+    def __init__(self, pr_calc: bool = True, rng: np.random.Generator = None):
```
Comment (EarlMilktea) [inline graphix/sim/base_backend.py:37]: Please use `| None` explicitly as specified in PEP.

File: graphix/sim/tensornet.py [hunk 11]
```diff
@@ -151,16 +152,32 @@ def measure(self, cmd: command.M):
...
+            vop = graphix.clifford.H.index
+            angle = -angle
+            s_domain, t_domain = t_domain, s_domain
+        elif cmd.plane == Plane.XZ:
+            vop = (graphix.clifford.H @ graphix.clifford.S).index
+            angle = -angle
+            t_domain = t_domain + s_domain
+            s_domain, t_domain = t_domain, s_domain
+        else:
+            assert False
```
Comment (EarlMilktea) [inline graphix/sim/tensornet.py:171]: Same as above.

File: tests/test_pattern.py [hunk 12]
```diff
@@ -1,24 +1,28 @@
...
-from graphix.command import M, N
+from graphix.command import CommandKind, E, M, N
 from graphix.pattern import CommandNode, Pattern
+from graphix.pauli import Plane
 from graphix.sim.density_matrix import DensityMatrix
 from graphix.sim.statevec import Statevec
 from graphix.simulator import PatternSimulator
 from graphix.transpiler import Circuit
 
-if TYPE_CHECKING:
```
Comment (EarlMilktea) [inline tests/test_pattern.py:21]: I personally prefer `TYPE_CHECKING`.

File: tests/test_pattern.py [hunk 13]
```diff
@@ -33,6 +37,17 @@ def compare_backend_result_with_statevec(backend: str, backend_state, statevec:
...
 
 
+Outcome = typing.Literal[0, 1]
+
+
+class IterGenerator:
+    def __init__(self, it: typing.Iterator[Outcome]) -> None:
+        self.__it = it
+
+    def choice(self, _outcomes) -> Outcome:
```
Comment (EarlMilktea) [inline tests/test_pattern.py:47]: May better to add `: Any`.

File: tests/test_pattern.py [hunk 14]
```diff
@@ -33,6 +37,17 @@ def compare_backend_result_with_statevec(backend: str, backend_state, statevec:
         raise NotImplementedError(backend)
 
 
+Outcome = typing.Literal[0, 1]
+
+
+class IterGenerator:
+    def __init__(self, it: typing.Iterator[Outcome]) -> None:
+        self.__it = it
```
Comment (EarlMilktea) [inline tests/test_pattern.py:45]: `Iterator` cannot be reused: this means that `IterGenerator` takes the ownership of `it` provided by caller.
Is it really what you wanted?

File: tests/test_pattern.py [hunk 15]
```diff
@@ -33,6 +37,17 @@ def compare_backend_result_with_statevec(backend: str, backend_state, statevec:
         raise NotImplementedError(backend)
 
 
+Outcome = typing.Literal[0, 1]
+
+
+class IterGenerator:
+    def __init__(self, it: typing.Iterator[Outcome]) -> None:
```
Comment (EarlMilktea) [inline tests/test_pattern.py:44]: Please import from `collections.abc`.

**Review by EarlMilktea — CHANGES_REQUESTED** [review]

**Comment by shinich1** [comment]
@thierry-martinez I think we can merge #182 before this, so that ruff will be added to checks (resolving that may auto-resolve some of above comments by @EarlMilktea).

File: graphix/command.py (hunk 1 above)
Comment (thierry-martinez) [reply graphix/command.py:11]: Fixed in f48ee93.

File: graphix/pattern.py (hunk 2 above)
Comment (thierry-martinez) [reply graphix/pattern.py:2196]: `signal` is the domain that the shift command will carry (and `s_domain` and `t_domain` are the new respective domain for the `M` command). I added some doc-comments in 4401e62 to clarify.

File: graphix/command.py (hunk 3 above)
Comment (thierry-martinez) [reply graphix/command.py:58]: Fixed in f48ee93.

File: graphix/command.py (hunk 4 above)
Comment (thierry-martinez) [reply graphix/command.py]: Fixed in adc6209.

File: graphix/command.py (hunk 5 above)
Comment (thierry-martinez) [reply graphix/command.py:70]: I don't think types would be currently precise enough to capture this: all the Clifford gates are now in a common class `Clifford`, and even if we introduce a new sub-class `IHSZ`, I don't know how to express the fact that the only instances would be `I`, `H`, `S` and `Z`. Perhaps a dedicated `Enum` would be a solution, but I would leave that for another PR.

File: graphix/command.py (hunk 6 above)
Comment (thierry-martinez) [reply graphix/command.py]: Fixed in `d055b75`.

File: graphix/sim/tensornet.py (hunk 11 above)
Comment (thierry-martinez) [reply graphix/sim/tensornet.py]: Fixed in f6f9d5a.

File: graphix/pattern.py (hunk 7 above)
Comment (thierry-martinez) [reply graphix/pattern.py]: Fixed in 01484dd.

File: tests/test_pattern.py (hunk 13 above)
Comment (thierry-martinez) [reply tests/test_pattern.py]: Fixed in 2312c49.

File: tests/test_pattern.py (hunk 14 above)
Comment (thierry-martinez) [reply tests/test_pattern.py]: I acknowledge it is safer to take an `Iterable` and use `iter` on it. Fixed in 41a9123.

File: tests/test_pattern.py (hunk 15 above)
Comment (thierry-martinez) [reply tests/test_pattern.py]: Fixed in 9cc65a5.

**Comment by EarlMilktea** [comment]
@thierry-martinez Could you resolve the remaining three threads? Once ready, I will approve merge.

File: graphix/sim/base_backend.py (hunk 10 above)
Comment (thierry-martinez) [reply graphix/sim/base_backend.py]: Fixed in dbd90dc.

File: graphix/pattern.py (hunk 8 above)
Comment (thierry-martinez) [reply graphix/pattern.py]: Fixed in b572ce5.

File: graphix/pattern.py (hunk 9 above)
Comment (thierry-martinez) [reply graphix/pattern.py]: Fixed in 6fce06f3.

**Comment by thierry-martinez** [comment]
Oops, sorry for having missed them (they were folded in "3 hidden more conversations")! All of them should be fixed now. Thanks!

**Review by EarlMilktea — APPROVED** [review]

**Comment by thierry-martinez** [comment]
Merged. Thanks!

**Pull request merged**